EntityManager doesn't work for quarkus scheduler

```
@ApplicationScoped
public class Scheduler {

  @Inject
  EntityManager entityManager;

  @Scheduled(every = "10s")
  void run() {
    Company company = entityManager.find(Company.class, 1);
    System.out.println(company);
  }
}
```


 ```
ERROR [org.qua.cor.JobRunShell] (DefaultQuartzScheduler_Worker-1) Job io.quarkus.scheduler.Scheduler.1_com.gateweb.transsexual.Scheduler_ScheduledInvoker_run_72e66771a77415a7284d3ae42331659c186071de threw an unhandled Exception: : javax.enterprise.context.ContextNotActiveException: interface javax.enterprise.context.RequestScoped
        at io.quarkus.hibernate.orm.runtime.RequestScopedEntityManagerHolder_ClientProxy.delegate(Unknown Source)
        at io.quarkus.hibernate.orm.runtime.RequestScopedEntityManagerHolder_ClientProxy.getOrCreateEntityManager(Unknown Source)
        at io.quarkus.hibernate.orm.runtime.entitymanager.TransactionScopedEntityManager.getEntityManager(TransactionScopedEntityManager.java:91)
        at io.quarkus.hibernate.orm.runtime.entitymanager.TransactionScopedEntityManager.find(TransactionScopedEntityManager.java:135)
        at io.quarkus.hibernate.orm.runtime.entitymanager.ForwardingEntityManager.find(ForwardingEntityManager.java:58)
        at com.gateweb.transsexual.Scheduler.run(Scheduler.java:18)
        at com.gateweb.transsexual.Scheduler_ClientProxy.run(Unknown Source)
        at com.gateweb.transsexual.Scheduler_ScheduledInvoker_run_72e66771a77415a7284d3ae42331659c186071de.invoke(Unknown Source)
        at io.quarkus.scheduler.runtime.QuartzScheduler$InvokerJob.execute(QuartzScheduler.java:264)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
```

```
 ERROR [org.qua.cor.ErrorLogger] (DefaultQuartzScheduler_Worker-1) Job (io.quarkus.scheduler.Scheduler.1_com.gateweb.transsexual.Scheduler_ScheduledInvoker_run_72e66771a77415a7284d3ae42331659c186071de threw an exception.: org.quartz.SchedulerException: Job threw an unhandled exception. [See nested exception: javax.enterprise.context.ContextNotActiveException: interface javax.enterprise.context.RequestScoped]
        at org.quartz.core.JobRunShell.run(JobRunShell.java:213)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: javax.enterprise.context.ContextNotActiveException: interface javax.enterprise.context.RequestScoped
        at io.quarkus.hibernate.orm.runtime.RequestScopedEntityManagerHolder_ClientProxy.delegate(Unknown Source)
        at io.quarkus.hibernate.orm.runtime.RequestScopedEntityManagerHolder_ClientProxy.getOrCreateEntityManager(Unknown Source)
        at io.quarkus.hibernate.orm.runtime.entitymanager.TransactionScopedEntityManager.getEntityManager(TransactionScopedEntityManager.java:91)
        at io.quarkus.hibernate.orm.runtime.entitymanager.TransactionScopedEntityManager.find(TransactionScopedEntityManager.java:135)
        at io.quarkus.hibernate.orm.runtime.entitymanager.ForwardingEntityManager.find(ForwardingEntityManager.java:58)
        at com.gateweb.transsexual.Scheduler.run(Scheduler.java:18)
        at com.gateweb.transsexual.Scheduler_ClientProxy.run(Unknown Source)
        at com.gateweb.transsexual.Scheduler_ScheduledInvoker_run_72e66771a77415a7284d3ae42331659c186071de.invoke(Unknown Source)
        at io.quarkus.scheduler.runtime.QuartzScheduler$InvokerJob.execute(QuartzScheduler.java:264)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        ... 1 more
```

--- a/extensions/scheduler/deployment/src/main/java/io/quarkus/scheduler/deployment/SchedulerProcessor.java
+++ b/extensions/scheduler/deployment/src/main/java/io/quarkus/scheduler/deployment/SchedulerProcessor.java
@@ -104,18 +104,26 @@ public class SchedulerProcessor {
 
             @Override
             public boolean appliesTo(org.jboss.jandex.AnnotationTarget.Kind kind) {
-                return kind == org.jboss.jandex.AnnotationTarget.Kind.CLASS;
+                return kind == org.jboss.jandex.AnnotationTarget.Kind.CLASS
+                        || kind == org.jboss.jandex.AnnotationTarget.Kind.METHOD;
             }
 
             @Override
             public void transform(TransformationContext context) {
-                if (context.getAnnotations().isEmpty()) {
+                if (context.isClass() && context.getAnnotations().isEmpty()) {
                     // Class with no annotations but with @Scheduled method
                     if (context.getTarget().asClass().annotations().containsKey(SCHEDULED_NAME)
                             || context.getTarget().asClass().annotations().containsKey(SCHEDULES_NAME)) {
                         LOGGER.debugf("Found scheduled business methods on a class %s with no annotations - adding @Singleton",
                                 context.getTarget());
                         context.transform().add(Singleton.class).done();
+                    }
+                } else if (context.isMethod()) {
+                    MethodInfo method = context.getTarget().asMethod();
+                    if ((method.hasAnnotation(SCHEDULED_NAME) || method.hasAnnotation(SCHEDULES_NAME))
+                            && !method.hasAnnotation(DotNames.ACTIVATE_REQUEST_CONTEXT)) {
+                        // Activate request context during a scheduled method invocation
+                        context.transform().add(DotNames.ACTIVATE_REQUEST_CONTEXT).done();
                     }
                 }
             }

--- a/extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/RequestContextJobs.java
+++ b/extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/RequestContextJobs.java
@@ -1,0 +1,55 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.scheduler.test;
+
+import java.util.concurrent.CountDownLatch;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.RequestScoped;
+import javax.inject.Inject;
+
+import io.quarkus.scheduler.Scheduled;
+
+public class RequestContextJobs {
+
+    static final CountDownLatch LATCH = new CountDownLatch(1);
+
+    @Inject
+    RequestFoo foo;
+
+    @Scheduled(every = "1s")
+    void checkEverySecond() {
+        foo.getName();
+        LATCH.countDown();
+    }
+
+    @RequestScoped
+    static class RequestFoo {
+
+        private String name;
+
+        @PostConstruct
+        void init() {
+            name = "oof";
+        }
+
+        public String getName() {
+            return name;
+        }
+
+    }
+
+}

--- a/extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/RequestScheduledMethodTest.java
+++ b/extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/RequestScheduledMethodTest.java
@@ -1,0 +1,40 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.quarkus.scheduler.test;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.concurrent.TimeUnit;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class RequestScheduledMethodTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class).addClasses(RequestContextJobs.class));
+
+    @Test
+    public void testRequestContextScheduledJobs() throws InterruptedException {
+        assertTrue(RequestContextJobs.LATCH.await(4, TimeUnit.SECONDS));
+    }
+
+}

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DotNames.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DotNames.java
@@ -21,6 +21,7 @@ import java.util.Optional;
 import javax.annotation.PostConstruct;
 import javax.annotation.PreDestroy;
 import javax.annotation.Priority;
+import javax.enterprise.context.control.ActivateRequestContext;
 import javax.enterprise.event.Event;
 import javax.enterprise.event.Observes;
 import javax.enterprise.event.ObservesAsync;
@@ -87,6 +88,7 @@ public final class DotNames {
     public static final DotName EXTENSION = create(Extension.class);
     public static final DotName OPTIONAL = create(Optional.class);
     public static final DotName NAMED = create(Named.class);
+    public static final DotName ACTIVATE_REQUEST_CONTEXT = create(ActivateRequestContext.class);
 
     public static final DotName BOOLEAN = create(Boolean.class);
     public static final DotName BYTE = create(Byte.class);

